SAT-34713 - Update Hybrid rake task to prompt for org_id and custom_url

chris1984 opened the pull request:
* foreman-rake was changed which now clears all env vars except TERM and --whitelist
* Adds new prompts to the user to ask for things we had in the env vars
* Added some comments, added better validation for inputs and added color to the prompts.
* Added some error handling so the ruby traceback is not ugly, with keeping the option to turn it back on(commented)

## Summary by Sourcery

Update the Hybrid Cloud Rake task to interactively prompt for organization ID, custom Insights URL, and API token instead of relying on environment variables, with input validation, colored output, and enhanced error handling.

Enhancements:
- Replace environment variable inputs with interactive prompts for organization ID, custom URL, and API token.
- Validate that the organization ID is numeric and that the organization and token exist before proceeding.
- Add colored console output and separators to improve prompt clarity and user experience.
- Warn when using the default Insights URL and support custom URL input.
- Refactor task into logical sections with helper methods and inline comments for better readability.
- Implement specific error handling for unauthorized and other HTTP errors with clear logging and exit behavior.

chris1984 committed:
SAT-34713 - Update Hybrid rake task to prompt for org_id and custom_url

* foreman-rake was changed here theforeman/foreman@3636aec, from helptext on runuser:
-, -l, --login
Start the shell as a login shell with an environment similar to a real login:
•clears all the environment variables except for TERM and variables specified by --whitelist-environment
•initializes the environment variables HOME, SHELL, USER, LOGNAME, and PATH
•changes to the target user’s home directory
•sets argv[0] of the shell to '-' in order to make the shell a login shell

* Adds new prompts to the user to ask for things we had in the env vars
* Added some comments, added better validation for inputs and added color to the prompts.
* Added some error handling so the ruby traceback is not ugly, with keeping the option to turn it back on(commented)

chris1984 commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,123 @@
+        ForemanRhCloud.base_url + '/api/identity/certificate/registrations'
+      else
+        logger.warn("Custom Insights URL is set to: #{custom_url}")
+        custom_url
```
Fixed

chris1984 commented:
@sourcery-ai review

chris1984 commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,123 @@
     include ::InsightsCloud::CandlepinCache
 
+    # Helper method to get a logger instance
     def logger
```
Fixed

chris1984 commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,123 @@
         headers: headers,
         payload: payload.to_json
       )
       logger.debug(response)
```
Fixed

jeremy-audet commented:
This commit message says:

> foreman-rake was changed which now clears all env vars except TERM and --whitelist

But the full help text from runuser(1) is:

> -, -l, --login
> Start the shell as a login shell with an environment similar to a real login:
> •clears all the environment variables except for TERM and variables specified by --whitelist-environment
> •initializes the environment variables HOME, SHELL, USER, LOGNAME, and PATH
> •changes to the target user’s home directory
> •sets argv[0] of the shell to '-' in order to make the shell a login shell

chris1984 commented:
> This commit message says:
> 
> > foreman-rake was changed which now clears all env vars except TERM and --whitelist
> 
> But the full help text from runuser(1) is:
> 
> > -, -l, --login
> > Start the shell as a login shell with an environment similar to a real login:
> > •clears all the environment variables except for TERM and variables specified by --whitelist-environment
> > •initializes the environment variables HOME, SHELL, USER, LOGNAME, and PATH
> > •changes to the target user’s home directory
> > •sets argv[0] of the shell to '-' in order to make the shell a login shell

Fixed the commit message to include the helptext you put into the Jira and comment.

jeremy-audet commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,124 @@
-      @logger ||= Logger.new(STDOUT)
+    # Helper method to get the registrations URL, with a warning for default usage
+    def registrations_url(custom_url = nil)
+      if custom_url.blank?
```
`nil.blank?` will throw an exception:

```
irb(main):001> "foo".empty?
=> false
irb(main):002> "".empty?
=> true
irb(main):003> nil.empty?
(irb):3:in `<main>': undefined method `empty?' for nil (NoMethodError)

nil.empty?
   ^^^^^^^
        from <internal:kernel>:187:in `loop'
        from /usr/share/gems/gems/irb-1.13.1/exe/irb:9:in `<top (required)>'
        from /usr/bin/irb:25:in `load'
        from /usr/bin/irb:25:in `<main>'
```

Can the function signature be changed to `registrations_url(custom_url = "")`?

chris1984 replied:
Set it to:
```ruby
def registrations_url(custom_url)
[1] pry(main)> custom_url
=> ""
[2] pry(main)> custom_url.empty?
=> true
```

It's coming back as `""` if I don't type anything and hit just hit enter.

jeremy-audet commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,124 @@
-      logger.warn("Custom url is not set, using the default one: #{ForemanRhCloud.base_url}") if ENV['SATELLITE_RH_CLOUD_URL'].empty?
-      ForemanRhCloud.base_url + '/api/identity/certificate/registrations'
+    # --- Input Collection ---
+    puts "\033[32mPaste in your organization ID, this can be retrieved with the command: hammer organization list\033[0m"
```
Is it a good idea to unconditionally print ANSI escape codes for setting font colors to the terminal? Not all terminals support colors.

chris1984 replied:
You make a good point, I will get rid of the colors and just keep the font normal.

jeremy-audet replied:
Color support is a really nice enhancement, as long as controls are provided. As a point of comparison, here's what git-log(1) says:

```
       --color[=<when>]
           Show colored diff.  --color (i.e. without =<when>) is the same as --color=always.
           <when> can be one of always, never, or auto.
```

But I've definitely encountered cases where escape codes were not supported, and them being printed made it hard to read the output I cared about.

chris1984 replied:
Removed the colors, for a script that is not used that much, no point to get fancy with it.

jeremy-audet commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,124 @@
+        logger.warn("Custom url is not set, using the default one: #{ForemanRhCloud.base_url}")
+        URI.join(ForemanRhCloud.base_url, '/api/identity/certificate/registrations').to_s
+      else
+        logger.warn("Custom Insights URL is set to: #{custom_url}")
```
If a user provides a custom URL, does it make sense to warn the user? My expectation is that a warning would only be provided in the case where a default URL is used.

chris1984 replied:
Removed the warning if using a custom URL

jeremy-audet commented:
This issue's title references SAT-34717, but I believe the actual issue is SAT-34713.

chris1984 commented:
> This issue's title references SAT-34717, but I believe the actual issue is SAT-34713.

Changed the commit message and PR title

jeremy-audet commented on lib/tasks/hybrid_cloud.rake:
```diff
@@ -1,64 +1,122 @@
-      }
+    uid = cp_owner_id(organization)
+    if uid.nil?
+      logger.error('Organization provided does not have a manifest imported.')
```
Could we add the organization name in this message? Something like:

```ruby
logger.error("Organization '#{organization}' does not have a manifest imported.")
```

chris1984 replied:
Fixed in our sync up

jeremy-audet commented:
Can we look at this together on a call? I have a bunch of suggestions that aren't _crucial_, but IMO would make this code nicer, and it would be way faster if we looked things over together.

jeremy-audet committed:
Refactor hybrid_cloud.rake

Do the following:

* Create methods to validate user input, and push token collection into
  the "input collection" section of the rake task. This lets the "data
  preparation" section of the script be more focused.
* Drop duplicate comments from the error handling block, for legibility.
* Print the default registration URL. This helps the user understand
  what the script is doing, and gives an example in the case where a
  user would like to provide a custom URL.
* Optionally prepend a scheme to a user-provided URL. That is,
  example.com will be transformed to https://example.com.
* Do not append a path to a user-provided custom URL. This ensures
  forward compatibility in the case that Insights requires registration
  at a new path (e.g. /api/identityv2/certificate/registrations).

(cherry picked from commit 914d2fb6d0ef02c28ca266d9a72a7ba59e3b53c6)

jeremy-audet reviewed (approved):
Tested against commercial, to the extent possible. I don't believe commercial supports mtls auth.

chris1984 commented:
Merging since Jeremy Audet tested and we did a pair program session to resolve his comments.
